affinity/cpp-23/dXXXXr0.md

```diff
@@ -0,0 +1,295 @@
+data.reserve(SIZE);
+
+// Create a NUMA-aware executor.
+numa_executor numaExec;
+
+// Create an iota view to represent the iteration space of the bulk algorithm.
+auto indexRng = ranges::iota_view{SIZE};
+
+// Create a new execution policy which requires the adjacency.constructive
+// property on the std::par execution policy so that it provides a hint to the
+// executor that the algorithm would benefit from constructive interference in
+// the mapping of work-items to execution resources.
+auto adjacencyPar = std::execution::require(std::par, adjacency.constructive);
+
+// Define a callable that will be invoked for each work-item in the bulk
+// algorithm and perform first touch initialization.
+auto initialize = [=](size_t idx, std::vector<unique_ptr<float>> &value) {
+  value[idx] = std::make_new<float>(0.0f);
+}
+
+// Define a callable that will be invoked for each work-item in the bulk
+// algorithm abd perform some computation.
+auto compute = [=](size_t idx, std::vector<unique_ptr<float>> &value) {
+    do_something(value[idx]);
+}
+
+// Compose a sender that takes the input data, schedules using the NUMA-aware
+// executor and then calls the indexed_for algorithm twice, once to initialize
+// and then again to perform the computation.
+auto sender = std::execution::just(data)
+            | std::execution::via(numaExec)
+            | std::execution::indexed_for(indexRng, adjacencyPar, initialize)
+            | std::execution::indexed_for(indexRng, adjacencyPar, compute);
+
+// Submit the sender and wait on the result.
+std::execution::sync_wait(sender, std::execution::sink_receiver{});
+```
+*Listing 1: Using the `adjacency properties on a NUMA-aware executor*
+
+### Execution resources
+
+As a pre-requisite for the `adjacency` properties, this paper also proposes a
+definition for the underlying resources of an executor.
+
+## Proposed wording
+
+### Synopsis
+
+```cpp
+namespace std {
+namespace experimental {
+namespace execution {
+
+struct adjacency_t {
+
+  struct no_implication_t;
+  struct constructive_t;
+  struct destructive_t;
+
+  constexpr no_implication_t no_implication;
+  constexpr constructive_t constructive;
+  constexpr destructive_t destructive;
+
+};
+
+constexpr adjacency_t adjacency;
+
+}  // execution
+}  // experimental
+}  // std
+```
+*Listing 2: Addition to execution header synopsis*
+
+### Execution resources
+
+An *execution resource* represents an abstraction of a hardware or software
+layer that guarantees a particular set of affinity properties, where the level
+of abstraction is implementation-defined. An implementation is permitted to
+migrate any underlying resources providing it guarantees the affinity properties
+remain consistent. This allows freedom for the implementor but also consistency
+for users.
+
+If an *execution resource* is valid, then it must always point to the same
+underlying thing. For example, a *resource* cannot first point to one CPU core,
```

Comment:
Given the VM example below, perhaps we should just delete the "For example..." sentence.  It might be a bit confusing.  The sentence after suffices, I think.

Reply:
I agree, I've removed this sentence.